File: src/semp_workflow/output.py
"""Ansible-style colored console output."""

from __future__ import annotations

import sys
from typing import TYPE_CHECKING

from colorama import Fore, Style, init as colorama_init

from .models import ResultStatus

if TYPE_CHECKING:
    from .models import ActionResult, WorkflowResult

# Initialize colorama for Windows support
colorama_init()

SEPARATOR = "=" * 70
TASK_SEP = "-" * 70


def print_banner() -> None:
    print(f"\n{Style.BRIGHT}{SEPARATOR}")
    print("  SEMP Workflow Automation")
    print(f"{SEPARATOR}{Style.RESET_ALL}\n")


def print_workflow_header(
    workflow_name: str, template_ref: str, inputs: dict, index: int = 0
) -> None:
    print(f"\n{Style.BRIGHT}PLAY {index} [{workflow_name}] ({template_ref}){Style.RESET_ALL}")
    if inputs:
        input_str = ", ".join(f"{k}={v}" for k, v in inputs.items())
        print(f"  Inputs: {input_str}")
    print(TASK_SEP)


def print_task_result(result: ActionResult) -> None:
    """Print a single task result with color-coded status."""
    status = result.status
    name = result.task_name or result.module

    if status == ResultStatus.OK:
        color = Fore.YELLOW
        label = "changed"
    elif status == ResultStatus.DRYRUN:
        color = Fore.CYAN
        label = "dryrun"
    elif status == ResultStatus.SKIPPED:
        color = Fore.CYAN
        label = "skipped"
    else:
        color = Fore.RED
        label = "FAILED"

    # Pad task name for alignment
    task_display = f"TASK [{name}]"
    padding = max(1, 55 - len(task_display))
    dots = "." * padding

    print(
        f"  {task_display} {dots} "
        f"{color}{Style.BRIGHT}{label}{Style.RESET_ALL}"
    )
    if result.message:
        msg_color = Fore.RED if status == ResultStatus.FAILED else ""
        print(f"    => {msg_color}{result.message}{Style.RESET_ALL}")


def print_dry_run_banner() -> None:
    print(f"\n{Fore.CYAN}{Style.BRIGHT}** DRY RUN MODE ** "
          f"No changes will be made{Style.RESET_ALL}")


def print_recap(results: list[WorkflowResult]) -> None:
    """Print the final recap summary."""
    print(f"\n{Style.BRIGHT}{SEPARATOR}")
    print(f"RECAP{Style.RESET_ALL}")
    print(SEPARATOR)

    total_ok = 0
    total_dryrun = 0
    total_skipped = 0
    total_failed = 0

    for i, wf in enumerate(results, 1):
        ok = wf.ok_count
        dryrun = wf.dryrun_count
        skipped = wf.skipped_count
        failed = wf.failed_count

        total_ok += ok
        total_dryrun += dryrun
        total_skipped += skipped
        total_failed += failed

        failed_str = (
            f"{Fore.RED}failed={failed}{Style.RESET_ALL}"
            if failed
            else f"failed={failed}"
        )

        print(
            f"  Workflow {i} ({wf.workflow_name}): "
            f"{Fore.YELLOW}changed={ok}{Style.RESET_ALL}  "
            f"{Fore.CYAN}dryrun={dryrun}{Style.RESET_ALL}  "
            f"{Fore.CYAN}skipped={skipped}{Style.RESET_ALL}  "
            f"{failed_str}"
        )

    print(TASK_SEP)
    total_failed_str = (
        f"{Fore.RED}{Style.BRIGHT}failed={total_failed}{Style.RESET_ALL}"
        if total_failed
        else f"failed={total_failed}"
    )
    print(
        f"  {Style.BRIGHT}Total: "
        f"{Style.BRIGHT}{Fore.YELLOW}changed={total_ok}{Style.RESET_ALL}  "
        f"{Style.BRIGHT}{Fore.CYAN}dryrun={total_dryrun}{Style.RESET_ALL}  "
        f"{Style.BRIGHT}{Fore.CYAN}skipped={total_skipped}{Style.RESET_ALL}  "
        f"{total_failed_str}"
    )
    print(SEPARATOR)

    if total_failed:
        print(f"\n{Fore.RED}{Style.BRIGHT}Some tasks failed!{Style.RESET_ALL}")
        sys.exit(1)
    else:
        print(f"\n{Fore.GREEN}{Style.BRIGHT}All tasks completed successfully.{Style.RESET_ALL}")


def print_module_list(modules: list[str]) -> None:
    """Print all available modules grouped by object."""
    print(f"\n{Style.BRIGHT}Available Modules:{Style.RESET_ALL}\n")

    # Group by object prefix
    groups: dict[str, list[str]] = {}
    for name in modules:
        obj, verb = name.split(".", 1)
        groups.setdefault(obj, []).append(verb)

    for obj in sorted(groups):
        print(f"  {Style.BRIGHT}{obj}{Style.RESET_ALL}")
        for verb in sorted(groups[obj]):
            print(f"    - {obj}.{verb}")
        print()


def print_validation_ok(config_path: str, template_count: int, workflow_count: int) -> None:
    print(f"\n{Fore.GREEN}{Style.BRIGHT}Validation passed!{Style.RESET_ALL}")
    print(f"  Config: {config_path}")
    print(f"  Templates loaded: {template_count}")
    print(f"  Workflows defined: {workflow_count}")


def print_error(message: str) -> None:
    print(f"\n{Fore.RED}{Style.BRIGHT}ERROR: {message}{Style.RESET_ALL}", file=sys.stderr)


def render_module_docs_md(module_info: dict[str, dict]) -> str:
    """Render all module metadata as a Markdown document."""
    lines: list[str] = []
    lines.append("# SEMP Workflow Automation — Module Reference")
    lines.append("")
    lines.append("All actions are **idempotent**: each checks current state before acting.")
    lines.append("Result states: `changed` (action ran), `skipped` (already in desired state), `dryrun` (would change), `failed` (error).")
    lines.append("")

    # Group by object prefix (queue, rdp, …)
    groups: dict[str, list[str]] = {}
    for name in module_info:
        obj, _ = name.split(".", 1)
        groups.setdefault(obj, []).append(name)

    # Table of contents
    lines.append("## Contents")
    lines.append("")
    for obj in sorted(groups):
        anchor = obj.replace("_", "-")
        lines.append(f"- [{obj}](#{anchor})")
    lines.append("")
    lines.append("---")
    lines.append("")

    for obj in sorted(groups):
        lines.append(f"## {obj}")
        lines.append("")

        for action_name in sorted(groups[obj]):
            info = module_info[action_name]
            description = info["description"]
            params = info["params"]

            lines.append(f"### `{action_name}`")
            lines.append("")
            if description:
                lines.append(description)
                lines.append("")

            if params:
                lines.append("| Parameter | Type | Required | Default | Description |")
                lines.append("|-----------|------|----------|---------|-------------|")
                for param_name, meta in params.items():
                    ptype    = meta.get("type", "string")
                    required = "Yes" if meta.get("required") else "No"
                    default  = f"`{meta['default']}`" if "default" in meta else "—"
                    desc     = meta.get("description", "")
                    if "enum" in meta:
                        allowed = ", ".join(f"`{v}`" for v in meta["enum"])
                        desc = f"{desc} ({allowed})" if desc else allowed
                    lines.append(f"| `{param_name}` | {ptype} | {required} | {default} | {desc} |")
                lines.append("")
            else:
                lines.append("_No parameters._")
                lines.append("")

        lines.append("---")
        lines.append("")

    return "\n".join(lines)

File: internal/output/output.go
// Package output renders Ansible-style colored console output and the module
// reference Markdown. Color is handled by fatih/color, which honors NO_COLOR and
// non-TTY output so captured logs stay clean.
package output

import (
	"fmt"
	"io"
	"os"
	"sort"
	"strings"

	"github.com/fatih/color"

	"semp-workflow/internal/models"
	"semp-workflow/internal/modules"
	"semp-workflow/internal/semp"
)

// Out and ErrOut are the destinations for normal and error output. They are
// package-level so tests can capture output; production uses stdout/stderr.
var (
	Out    io.Writer = os.Stdout
	ErrOut io.Writer = os.Stderr
)

// SEPARATOR and TASK_SEP are the horizontal rules used in the output.
var (
	SEPARATOR = strings.Repeat("=", 70)
	TASK_SEP  = strings.Repeat("-", 70)
)

var (
	cBold   = color.New(color.Bold)
	cYellow = color.New(color.FgYellow, color.Bold)
	cCyan   = color.New(color.FgCyan, color.Bold)
	cRed    = color.New(color.FgRed, color.Bold)
	cGreen  = color.New(color.FgGreen, color.Bold)
)

// PrintBanner prints the top-of-run banner.
func PrintBanner() {
	fmt.Fprintf(Out, "\n%s\n", cBold.Sprint(SEPARATOR))
	fmt.Fprintln(Out, "  SEMP Workflow Automation")
	fmt.Fprintf(Out, "%s\n\n", cBold.Sprint(SEPARATOR))
}

// PrintWorkflowHeader prints the PLAY header for a workflow.
func PrintWorkflowHeader(workflowName, templateRef string, inputs map[string]any, index int) {
	fmt.Fprintf(Out, "\n%s\n", cBold.Sprintf("PLAY %d [%s] (%s)", index, workflowName, templateRef))
	if len(inputs) > 0 {
		keys := make([]string, 0, len(inputs))
		for k := range inputs {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			parts = append(parts, fmt.Sprintf("%s=%s", k, semp.Stringify(inputs[k])))
		}
		fmt.Fprintf(Out, "  Inputs: %s\n", strings.Join(parts, ", "))
	}
	fmt.Fprintln(Out, TASK_SEP)
}

// PrintTaskResult prints a single task result with a color-coded status.
func PrintTaskResult(result models.ActionResult) {
	name := result.TaskName
	if name == "" {
		name = result.Module
	}

	var col *color.Color
	var label string
	switch result.Status {
	case models.StatusOK:
		col, label = cYellow, "changed"
	case models.StatusDryRun:
		col, label = cCyan, "dryrun"
	case models.StatusSkipped:
		col, label = cCyan, "skipped"
	default:
		col, label = cRed, "FAILED"
	}

	taskDisplay := fmt.Sprintf("TASK [%s]", name)
	padding := 55 - len(taskDisplay)
	if padding < 1 {
		padding = 1
	}
	dots := strings.Repeat(".", padding)

	fmt.Fprintf(Out, "  %s %s %s\n", taskDisplay, dots, col.Sprint(label))
	if result.Message != "" {
		if result.Status == models.StatusFailed {
			fmt.Fprintf(Out, "    => %s\n", color.New(color.FgRed).Sprint(result.Message))
		} else {
			fmt.Fprintf(Out, "    => %s\n", result.Message)
		}
	}
}

// PrintDryRunBanner announces dry-run mode.
func PrintDryRunBanner() {
	fmt.Fprintf(Out, "\n%s\n", cCyan.Sprint("** DRY RUN MODE ** No changes will be made"))
}

// PrintRecap prints the final recap summary and reports whether any task failed.
func PrintRecap(results []models.WorkflowResult) bool {
	fmt.Fprintf(Out, "\n%s\n", cBold.Sprint(SEPARATOR))
	fmt.Fprintln(Out, cBold.Sprint("RECAP"))
	fmt.Fprintln(Out, SEPARATOR)

	totalOK, totalDryRun, totalSkipped, totalFailed := 0, 0, 0, 0

	for i := range results {
		wf := &results[i]
		okc := wf.OKCount()
		dryc := wf.DryRunCount()
		skpc := wf.SkippedCount()
		fldc := wf.FailedCount()

		totalOK += okc
		totalDryRun += dryc
		totalSkipped += skpc
		totalFailed += fldc

		failedStr := fmt.Sprintf("failed=%d", fldc)
		if fldc > 0 {
			failedStr = cRed.Sprintf("failed=%d", fldc)
		}

		fmt.Fprintf(Out, "  Workflow %d (%s): %s  %s  %s  %s\n",
			i+1, wf.WorkflowName,
			cYellow.Sprintf("changed=%d", okc),
			cCyan.Sprintf("dryrun=%d", dryc),
			cCyan.Sprintf("skipped=%d", skpc),
			failedStr,
		)
	}

	fmt.Fprintln(Out, TASK_SEP)
	totalFailedStr := fmt.Sprintf("failed=%d", totalFailed)
	if totalFailed > 0 {
		totalFailedStr = cRed.Sprintf("failed=%d", totalFailed)
	}
	fmt.Fprintf(Out, "  %s %s  %s  %s  %s\n",
		cBold.Sprint("Total:"),
		cYellow.Sprintf("changed=%d", totalOK),
		cCyan.Sprintf("dryrun=%d", totalDryRun),
		cCyan.Sprintf("skipped=%d", totalSkipped),
		totalFailedStr,
	)
	fmt.Fprintln(Out, SEPARATOR)

	if totalFailed > 0 {
		fmt.Fprintf(Out, "\n%s\n", cRed.Sprint("Some tasks failed!"))
		return true
	}
	fmt.Fprintf(Out, "\n%s\n", cGreen.Sprint("All tasks completed successfully."))
	return false
}

// PrintModuleList prints all available modules grouped by object prefix.
func PrintModuleList(names []string) {
	fmt.Fprintf(Out, "\n%s\n\n", cBold.Sprint("Available Modules:"))

	groups := map[string][]string{}
	for _, name := range names {
		obj, verb := splitAction(name)
		groups[obj] = append(groups[obj], verb)
	}

	for _, obj := range sortedKeys(groups) {
		fmt.Fprintf(Out, "  %s\n", cBold.Sprint(obj))
		verbs := groups[obj]
		sort.Strings(verbs)
		for _, verb := range verbs {
			fmt.Fprintf(Out, "    - %s.%s\n", obj, verb)
		}
		fmt.Fprintln(Out)
	}
}

// PrintValidationOK prints the validation success summary.
func PrintValidationOK(configPath string, templateCount, workflowCount int) {
	fmt.Fprintf(Out, "\n%s\n", cGreen.Sprint("Validation passed!"))
	fmt.Fprintf(Out, "  Config: %s\n", configPath)
	fmt.Fprintf(Out, "  Templates loaded: %d\n", templateCount)
	fmt.Fprintf(Out, "  Workflows defined: %d\n", workflowCount)
}

// PrintError prints an error message to stderr.
func PrintError(message string) {
	fmt.Fprintf(ErrOut, "\n%s\n", cRed.Sprintf("ERROR: %s", message))
}

// RenderModuleDocsMD renders all module metadata as a Markdown document.
func RenderModuleDocsMD(moduleInfo map[string]modules.ModuleInfo) string {
	var lines []string
	add := func(s string) { lines = append(lines, s) }

	add("# SEMP Workflow Automation — Module Reference")
	add("")
	add("All actions are **idempotent**: each checks current state before acting.")
	add("Result states: `changed` (action ran), `skipped` (already in desired state), `dryrun` (would change), `failed` (error).")
	add("")

	// Group by object prefix (queue, rdp, …).
	groups := map[string][]string{}
	for name := range moduleInfo {
		obj, _ := splitAction(name)
		groups[obj] = append(groups[obj], name)
	}

	// Table of contents.
	add("## Contents")
	add("")
	for _, obj := range sortedKeys(groups) {
		anchor := strings.ReplaceAll(obj, "_", "-")
		add(fmt.Sprintf("- [%s](#%s)", obj, anchor))
	}
	add("")
	add("---")
	add("")

	for _, obj := range sortedKeys(groups) {
		add(fmt.Sprintf("## %s", obj))
		add("")

		actionNames := groups[obj]
		sort.Strings(actionNames)
		for _, actionName := range actionNames {
			info := moduleInfo[actionName]

			add(fmt.Sprintf("### `%s`", actionName))
			add("")
			if info.Description != "" {
				add(info.Description)
				add("")
			}

			if len(info.Params) > 0 {
				add("| Parameter | Type | Required | Default | Description |")
				add("|-----------|------|----------|---------|-------------|")
				for _, p := range info.Params {
					ptype := p.Type
					if ptype == "" {
						ptype = "string"
					}
					required := "No"
					if p.Required {
						required = "Yes"
					}
					def := "—"
					if p.Default != nil {
						def = fmt.Sprintf("`%s`", semp.Stringify(p.Default))
					}
					desc := p.Description
					if len(p.Enum) > 0 {
						allowed := make([]string, len(p.Enum))
						for i, v := range p.Enum {
							allowed[i] = fmt.Sprintf("`%s`", v)
						}
						joined := strings.Join(allowed, ", ")
						if desc != "" {
							desc = fmt.Sprintf("%s (%s)", desc, joined)
						} else {
							desc = joined
						}
					}
					add(fmt.Sprintf("| `%s` | %s | %s | %s | %s |", p.Name, ptype, required, def, desc))
				}
				add("")
			} else {
				add("_No parameters._")
				add("")
			}
		}

		add("---")
		add("")
	}

	return strings.Join(lines, "\n")
}

// splitAction splits an "object.verb" action name into its two parts.
func splitAction(name string) (obj, verb string) {
	parts := strings.SplitN(name, ".", 2)
	obj = parts[0]
	if len(parts) > 1 {
		verb = parts[1]
	}
	return obj, verb
}

// sortedKeys returns the sorted keys of a group map.
func sortedKeys(groups map[string][]string) []string {
	keys := make([]string, 0, len(groups))
	for k := range groups {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}
